### PR TITLE
BCF-2854 Jfpc pipeline cache

### PR DESCRIPTION
**Pull request by ilija42**
Seems like the least invasive way of doing this. Putting this into libocr makes it hard to configure the cache

### CODE REVIEW

File: core/services/ocr2/plugins/median/config/config.go [hunk 1]
```diff
@@ -4,14 +4,18 @@
...
 	"github.com/pkg/errors"
 
 	"github.com/smartcontractkit/chainlink/v2/core/services/pipeline"
 )
 
 // The PluginConfig struct contains the custom arguments needed for the Median plugin.
 type PluginConfig struct {
-	JuelsPerFeeCoinPipeline string `json:"juelsPerFeeCoinSource"`
+	JuelsPerFeeCoinPipeline      string        `json:"juelsPerFeeCoinSource"`
+	JuelsPerFeeCoinCaching       bool          `json:"juelsPerFeeCoinCaching"`
```
Comment (krehermann) [inline core/services/ocr2/plugins/median/config/config.go:17]: i'd get rid of this and infer the behavior from a non-zero cache duration

File: core/services/ocr2/plugins/median/services.go [hunk 2]
```diff
@@ -113,18 +113,30 @@ func NewMedianServices(ctx context.Context,
...
+
+	juelsPerFeeCoinSource := ocrcommon.NewInMemoryDataSource(pipelineRunner, jb, pipeline.Spec{
 		ID:           jb.ID,
 		DotDagSource: pluginConfig.JuelsPerFeeCoinPipeline,
 		CreatedAt:    time.Now(),
 	}, lggr)
 
+	if pluginConfig.JuelsPerFeeCoinCaching {
+		lggr.Infof("juelsPerFeeCoinSource caching is enabled for jb %s, cache expiration is set to %s", jb.Name.ValueOrZero(), pluginConfig.JuelsPerFeeCoinCacheDuration.String())
+		juelsPerFeeCoinSource = ocrcommon.NewInMemoryDataSourceCache(pipelineRunner, jb, pipeline.Spec{
```
Comment (krehermann) [inline core/services/ocr2/plugins/median/services.go:131]: why does the cache need such a complex signature? instead of wrapping the pipeline i think would either change the datasource to take an optional caching configuration or i would have wrapped the datasource directly ~

juelsPer... = NewCachableDS(juelPer...)
or
juelsPer.. = NewInMemoryDataSource(a, b, cfg)

cfg = struct {
x
y
cacheOpts
}

File: core/services/ocrcommon/data_source.go [hunk 3]
```diff
@@ -176,19 +194,94 @@ func (ds *inMemoryDataSource) parse(finalResult pipeline.FinalResult) (*big.Int,
...
+	cacheExpiration time.Duration
+	mu              *sync.RWMutex
+	lastUpdate      atomic.Pointer[time.Time]
+	latestTrrs      pipeline.TaskRunResults
+	latestResult    pipeline.FinalResult
+	*inMemoryDataSource
+}
+
+// isUpdateOverdue atomically checks if update is overdue.
+func (ds *inMemoryDataSourceCache) isUpdateOverdue() bool {
```
Comment (krehermann) [inline core/services/ocrcommon/data_source.go:224]: this seems like unneeded in direction. the updater is on a ticker set by the cache TTL

won't the answer to this isUpdateOverdue always be yes?

File: core/services/ocrcommon/data_source.go [hunk 4]
```diff
@@ -176,19 +194,94 @@ func (ds *inMemoryDataSource) parse(finalResult pipeline.FinalResult) (*big.Int,
...
+		ds.latestTrrs = trrs
+		ds.lastUpdate.Store(&lastUpdate)
+	}
+	return nil
+}
+
+func (ds *inMemoryDataSourceCache) Observe(ctx context.Context, timestamp ocr2types.ReportTimestamp) (*big.Int, error) {
+	// if cache update is overdue revert to standard inMemoryDataSource Observe behaviour
+	result := make(chan error)
+	go func() { result <- ds.updateCache(ctx) }()
```
Comment (krehermann) [inline core/services/ocrcommon/data_source.go:262]: whats going on here? it seems like you avhe 2 routines updating the cache? one here and one via `updater`?

i'd update the cache in one routine and just read the value of the cache here

File: core/services/ocrcommon/data_source.go [hunk 5]
```diff
@@ -176,19 +194,94 @@ func (ds *inMemoryDataSource) parse(finalResult pipeline.FinalResult) (*big.Int,
...
+	select {
+	case <-ctx.Done():
+	case err := <-result:
+		if err != nil {
+			ds.lggr.Warnw("unable to update in memory data source cache during observe, returning most recent value", "err", err)
+		}
+	}
+
+	ds.mu.RLock()
```
Comment (krehermann) [inline core/services/ocrcommon/data_source.go:271]: if you have take the above suggestion, then the read lock can go inside of get() and this will be cleaner

File: core/services/ocrcommon/data_source.go (hunk 4 above)
Comment (ilija42) [reply core/services/ocrcommon/data_source.go:262]: The idea was to minimize outdated values, there are two possibilities here:

1. `updater` successfully updated cache and `updateCache` just immediately returns from this go routine
2. `updater` failed to update cache and this new go routine that is spawned in `Observe` updates cache before returning values

Comment (ilija42) [reply core/services/ocrcommon/data_source.go:262]: I can do the same thing but hide it behind cache get

File: core/services/ocrcommon/data_source.go (hunk 3 above)
Comment (krehermann) [reply core/services/ocrcommon/data_source.go:224]: i'd try to refactor to ~
```
type cache {
	mu sync.Mutex
	vals your_data // lastest data
	err error // lastest error if any

	ds datasource
}

func( c *cache) get() ( vals, error) {
	
	c.mu.rlock()
	err :=  c.err
	c.mu.runlock()
	
	if err != nil {
	// cache is stale, try manual update
	c.update()
}
	c.mu.rlock()
defer c.mu.runlock()
	return c.vals, c.err
}
// synchoronous update to the cache value
func (c *cache) update() {
	c.mu.lock()
	defer c.mu.unlock()
	// use datasource to compute a new value
}

// runs update in polling loop
func (c *cache) poll(ctx) {
	ticker = blah
	
	switch {
	ticker: c.update()
	ctx.Done: return
	}
	}

File: core/services/ocrcommon/data_source.go (hunk 4 above)
Comment (EasterTheBunny) [inline core/services/ocrcommon/data_source.go:262]: It looks like you might end up with lock contention on the `rs.mu` mutex. There is a `Lock` in `updateCache` and a subsequent `RLock` in `executeRun` for `currentAnswer`.

The docs say about mutexes: `It should not be used for recursive read locking; a blocked Lock call excludes new readers from acquiring the lock. See the documentation on the RWMutex type.`

**Review by EasterTheBunny — CHANGES_REQUESTED** [review]

File: core/services/ocrcommon/data_source.go (hunk 4 above)
Comment (ilija42) [reply core/services/ocrcommon/data_source.go:262]: Its not an issue, it will just wait till the update is done and have the freshest data. In real world its almost imossible to happen as it either updated the cache right before or the cache was already up to date, which means that this can only happen if there was an update due in 1 milisecond and the lock got grabbed right before RLock() locked the read. 
tldr There are no race conditions, just a small chance of observe waiting for a fresh update which is how observe works currently for juels fee per coin anyway

Comment (ilija42) [reply core/services/ocrcommon/data_source.go:262]: This is a readLock because I am looking if there is an easy way to make this cache work for the whole node instead of per job

File: core/services/ocrcommon/data_source.go (hunk 5 above)
Comment (ilija42) [reply core/services/ocrcommon/data_source.go:264]: Still has to have read lock to not cause a race condition complaint when filling out EATelemetry

File: core/services/ocrcommon/data_source.go [hunk 6]
```diff
@@ -176,19 +199,90 @@ func (ds *inMemoryDataSource) parse(finalResult pipeline.FinalResult) (*big.Int,
...
+
 	return ds.parse(finalResult)
 }
 
+// inMemoryDataSourceCache is a time based cache wrapper for inMemoryDataSource.
+// If cache update is overdue Observe defaults to standard inMemoryDataSource behaviour.
+type inMemoryDataSourceCache struct {
+	*inMemoryDataSource
+	cacheExpiration time.Duration
+	mu              *sync.RWMutex
```
Comment (krehermann) [inline core/services/ocrcommon/data_source.go:222]: this doesn't need to be a pointer. a zero value works without initialization

`mu sync.RWMutex` and then you can remove the explicit initialization in other call sites. i.e. omit setting mu when calling &inMemoryDataSourceCache{..}

`The zero value for a RWMutex is an unlocked mutex.` https://pkg.go.dev/sync#RWMutex

**Review by krehermann — DISMISSED** [review]

File: core/services/ocrcommon/data_source.go (hunk 6 above)
Comment (jmank88) [reply core/services/ocrcommon/data_source.go]: You can always make a pointer to the value, and the compiler even does that for you automatically when you call methods with pointer receivers.

**Review by chudilka1 — COMMENTED** [review]
Please, add the description to a CHANGELOG with a bit of details about the default values and the time range allowed to be set for `JuelsPerFeeCoinCacheDuration`

**Review by krehermann — DISMISSED** [review]

**Review by krehermann — APPROVED** [review]

**Review by EasterTheBunny — APPROVED** [review]
